Source/Task/AsyncLib.cpp

```diff
@@ -189,7 +292,17 @@ static HRESULT AllocState(_In_ AsyncBlock* asyncBlock)
     // If the async block is already associated with another
     // call, fail.
 
-    RETURN_HR_IF(E_INVALIDARG, GetState(asyncBlock) != nullptr);
+    // There is no great way to tell if the AsyncBlockInternal has already been
+    // initialized, because uninitialized memory can look like anything.
+    // Here we rely on the client zeoring out the entirety of the AsyncBlock
+    // object so we can check that AsyncBlock::Internal is all 0
+    for (auto i = 0u; i < sizeof(asyncBlock->internal); ++i)
+    {
+        if (asyncBlock->internal[i] != 0)
+        {
+            return E_INVALIDARG;
+        }
+    }
 
     HRESULT hr = AllocStateNoCompletion(asyncBlock);
```

Comment:
If this guy fails the failed HR needs to become the status of the async block.  I would make the AsyncBlockInternal connection two phased:  early in AllocStateNoCompletion, establish AsyncBlockInternal by blitting it into the internal fields and it should have a null state pointer.  That way you have enough established that you can set the HRESULT.  Completion callbacks are going to call GetAsyncStatus and will need that HR.